db/log_reader.cc

```diff
@@ -244,36 +284,43 @@ void Reader::ReportDrop(size_t bytes, const Status& reason) {
   }
 }
 
-unsigned int Reader::ReadPhysicalRecord(Slice* result,
-                                        const bool report_eof_inconsistency) {
+unsigned int Reader::ReadMore(size_t* drop_size) {
+  if (!eof_ && !read_error_) {
+    // Last read was a full read, so this is a trailer to skip
+    buffer_.clear();
+    Status status = file_->Read(kBlockSize, &buffer_, backing_store_);
+    end_of_buffer_offset_ += buffer_.size();
+    if (!status.ok()) {
+      buffer_.clear();
+      ReportDrop(kBlockSize, status);
+      read_error_ = true;
+      return kEof;
+    } else if (buffer_.size() < (size_t)kBlockSize) {
+      eof_ = true;
+      eof_offset_ = buffer_.size();
+    }
+    return 0;
+  } else {
+    // Note that if buffer_ is non-empty, we have a truncated header at the
+    //  end of the file, which can be caused by the writer crashing in the
+    //  middle of writing the header. Unless explicitly requested we don't
+    //  considering this an error, just report EOF.
+    if (buffer_.size()) {
+      *drop_size = buffer_.size();
+      buffer_.clear();
+      return kBadHeader;
+    }
+    buffer_.clear();
+    return kEof;
+  }
+}
+
+unsigned int Reader::ReadPhysicalRecord(Slice* result, size_t* drop_size) {
   while (true) {
+    // We need at least the minimum header size
     if (buffer_.size() < (size_t)kHeaderSize) {
-      if (!eof_ && !read_error_) {
-        // Last read was a full read, so this is a trailer to skip
-        buffer_.clear();
-        Status status = file_->Read(kBlockSize, &buffer_, backing_store_);
-        end_of_buffer_offset_ += buffer_.size();
-        if (!status.ok()) {
-          buffer_.clear();
-          ReportDrop(kBlockSize, status);
-          read_error_ = true;
-          return kEof;
-        } else if (buffer_.size() < (size_t)kBlockSize) {
-          eof_ = true;
-          eof_offset_ = buffer_.size();
-        }
-        continue;
-      } else {
-        // Note that if buffer_ is non-empty, we have a truncated header at the
-        //  end of the file, which can be caused by the writer crashing in the
-        //  middle of writing the header. Unless explicitly requested we don't
-        //  considering this an error, just report EOF.
-        if (buffer_.size() && report_eof_inconsistency) {
-          ReportCorruption(buffer_.size(), "truncated header");
-        }
-        buffer_.clear();
-        return kEof;
-      }
+      unsigned int r = ReadMore(drop_size);
+      if (r) return r;
```

Comment:
Can you change the signature. kEof if enum type and 0 is value, there is an implicit assumption that kEof != 0